sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala

```diff
@@ -0,0 +1,156 @@
+
+  case class BriefContactWithDataPartitionColumn(name: Name, address: String, p: Int)
+
+  val contactsWithDataPartitionColumn =
+    contacts.map { case Contact(name, address, pets, friends, relatives) =>
+      ContactWithDataPartitionColumn(name, address, pets, friends, relatives, 1) }
+  val briefContactsWithDataPartitionColumn =
+    briefContacts.map { case BriefContact(name: Name, address: String) =>
+      BriefContactWithDataPartitionColumn(name, address, 2) }
+
+  testSchemaPruning("select a single complex field") {
+    val query = sql("select name.middle from contacts")
+    checkScanSchemata(query, "struct<name:struct<middle:string>>")
+    checkAnswer(query, Row("X.") :: Row("Y.") :: Row(null) :: Row(null) :: Nil)
+  }
+
+  testSchemaPruning("select a single complex field and the partition column") {
+    val query = sql("select name.middle, p from contacts")
+    checkScanSchemata(query, "struct<name:struct<middle:string>>")
+    checkAnswer(query, Row("X.", 1) :: Row("Y.", 1) :: Row(null, 2) :: Row(null, 2) :: Nil)
+  }
+
+  testSchemaPruning("partial schema intersection - select missing subfield") {
+    val query = sql("select name.middle, address from contacts where p=2")
+    checkScanSchemata(query, "struct<name:struct<middle:string>,address:string>")
+    checkAnswer(query,
+      Row(null, "567 Maple Drive") ::
+      Row(null, "6242 Ash Street") :: Nil)
+  }
+
+  testSchemaPruning("no unnecessary schema pruning") {
+    val query =
+      sql("select name.last, name.middle, name.first, relatives[''].last, " +
+        "relatives[''].middle, relatives[''].first, friends[0].last, friends[0].middle, " +
+        "friends[0].first, pets, address from contacts where p=2")
+    // We've selected every field in the schema. Therefore, no schema pruning should be performed.
+    // We check this by asserting that the scanned schema of the query is identical to the schema
+    // of the contacts relation, even though the fields are selected in different orders.
+    checkScanSchemata(query,
+      "struct<name:struct<first:string,middle:string,last:string>,address:string,pets:int," +
+      "friends:array<struct<first:string,middle:string,last:string>>," +
+      "relatives:map<string,struct<first:string,middle:string,last:string>>>")
+    checkAnswer(query,
+      Row("Jones", null, "Janet", null, null, null, null, null, null, null, "567 Maple Drive") ::
+      Row("Jones", null, "Jim", null, null, null, null, null, null, null, "6242 Ash Street") ::
+      Nil)
+  }
+
+  testSchemaPruning("empty schema intersection") {
+    val query = sql("select name.middle from contacts where p=2")
+    checkScanSchemata(query, "struct<name:struct<middle:string>>")
+    checkAnswer(query,
+      Row(null) :: Row(null) :: Nil)
+  }
+
+  private def testSchemaPruning(testName: String)(testThunk: => Unit) {
+    withSQLConf(SQLConf.PARQUET_WRITE_LEGACY_FORMAT.key -> "false") {
```

Comment:
Whoa. I just noticed I've been switching on/off the wrong configuration option in my tests. I've meant to be testing with and without the vectorized reader. So I should be using `PARQUET_VECTORIZED_READER_ENABLED`, right?

Also, just to make sure, how many parquet reader implementations does Spark have? I think it's two. There's the one from the parquet-mr library, and there's another (vectorized) one that's part of the Spark codebase. If I'm wrong about that please alert me, because I'm making some erroneous assumptions.

Reply:
Yup.

If I am not mistaken, Spark has its's own implementation for vectorized reader alone, and we have a normal Parquet reader otherwise.